### ISSUE TITLE
Image size for drawimage$2

### ISSUE THREAD

**Issue by gagern** [CindyJS/CindyJS#466]
The function `drawimage(‹pos›,‹name›)` behaves differently in Cinderella and CindyJS. In the former, the size of the image is computed with respect to the user coordinate system, so that zooming in to the construction will zoom into the image. The latter renders the image at its natural size, so that zooming (which is not possible dynamically via the current UI anyway) will not change the apparent size. Both of these have benefits and drawbacks. And while compatibility with Cinderella mandates one behavior, compatibility with existing CindyJS content mandates another.

So which of these compatibility expectations do we prefer to break? As @richter-gebert has probably the best overview over who is using what feature of CindyScript, I'd value his input here.

**Comment by kortenkamp** [comment]
It is absolutely necessary to keep that behavior as is. I know many Cinderella examples where this is used. 

Ulli

**Comment by gagern** [comment]
In that case you *don't* want to keep it as it is *in CindyJS*, which means you have to adjust CindyJS to match Cinderella. I'll have a pull request for this in a moment.

**Comment by richter-gebert** [comment]
Yes I Agree with Ulli. Do I see it right that this only affects the situation of having exactly one reference point? If so, there should be little to no examples making use of this. Perhaps some UI icons for the exhibition apps are affected.

One other important point on the whish list: so far cindyjs is not capable of handling non- standard reference point locations. It would be great to have this.

Another one: how about making the 4 reference point case part of the cindygl plugin. As well as drawimagemap.  The latter only supporting those functions that are compatible with Aaron's transpiler.

Cu

Jürgen

Von meinem iPhone gesendet

**Comment by gagern** [comment]
Thanks @richter-gebert for the feedback!

Yes, this is only for the case of a single reference point. The exhibition widgets can stay on the 0.7 branch to ensure compatibility. I intend to maintain that branch for the foreseeable future. I'll wait for a review on #467.

I too noticed that the `ref` modifier doesn't work as it does in Cinderella. In fact I reported this as #34 quite some time ago, and just assigned that ticket to myself. I'll take care of this.

I'm all for using WebGL to render the projectively transformed images. I'm not sure if that's best handled using CindyGL, or a few lines of WebGL code inside CindyJS core. On the one hand it's core functionality in Cinderella, and shouldn't be too hard to do. On the other hand, we have made good experiences with auto-loading plugins (like quickhull3d) if certain functions are being used, so the plugin requirement might even go unnoticed by users. So I'd say @montaga should decide where he'd prefer to have this.

For the other, I think you meant [`mapimage`](https://github.com/CindyJS/CindyJS/blob/v0.7.3/ref/Image_Manipulation_and_Rendering.md#deforming-an-image-mapimageimagenamefunction). I'm not sure how well this can fit in with how CindyGL operates. As far as I can tell, the function there is applied in the forward direction, i.e. taking picture-space coordinates and transforming them to display-space. CindyGL would essentially require the inverse: from display-space to picture-space. Automatically inverting arbitrary functions sounds like a *very* tricky task. So I guess we'd be back to forward-sampling on a grid and then interpolating between these samples, which is more like a regular triangulate-and-texture rendering than the one-quad-only rendering CindyGL uses. So again @montaga should decide where to place this.

By the way, I noticed that the CindyScript documentation is *very* confusing here. [The page I referenced](https://github.com/CindyJS/CindyJS/blob/v0.7.3/ref/Image_Manipulation_and_Rendering.md#deforming-an-image-mapimageimagenamefunction) shows a picture getting slimmed down in the middle, while the *x* coordinate is modulated by sin(*y*) and the range for *y* is [0,π]. I tried for a long time to understand how this fits in with the examples from the [`mapgrid`](https://github.com/CindyJS/CindyJS/blob/v0.7.3/ref/Function_Plotting.md#mapping-a-rectangular-grid-mapgridexpr) documentation before I actually tried out the code and found that the image will get fattened, not slimmed. Thus confirming it's a forward transformation.
